### PR TITLE
Optimize: faster shuffling with a single per-epoch cached committee shuffle 

### PR DESCRIPTION
**Pull request by Nashatyrev**
# PR Description

Shuffling indexes per epoch in a single list pass is much faster than computing their permutations individually.

#### Before
```
Benchmark                              (validatorsCount)  Mode  Cnt  Score   Error  Units
TransitionBenchmark.Block.importBlock              16384    ss   50  0,266 ± 0,005   s/op
TransitionBenchmark.Block.importBlock              32768    ss   50  0,387 ± 0,004   s/op
TransitionBenchmark.Epoch.importBlock              16384    ss   10  0,897 ± 0,042   s/op
TransitionBenchmark.Epoch.importBlock              32768    ss   10  1,638 ± 0,018   s/op
```
#### After
```
Benchmark                              (validatorsCount)  Mode  Cnt  Score   Error  Units
TransitionBenchmark.Block.importBlock              16384    ss   50  0,183 ± 0,004   s/op
TransitionBenchmark.Block.importBlock              32768    ss   50  0,219 ± 0,006   s/op
TransitionBenchmark.Epoch.importBlock              16384    ss   10  0,805 ± 0,018   s/op
TransitionBenchmark.Epoch.importBlock              32768    ss   10  1,493 ± 0,033   s/op
```

### CODE REVIEW

File: eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.tuweni.crypto.Hash;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+
+@State(Scope.Thread)
+public class Sha256Benchmark {
```
Comment (ajsutton) [inline eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java:29]: Out of interest, do you have any idea on how our Sha256 performance compares to other implementations (particularly native ones)?

File: ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.stream.IntStream;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.jupiter.api.Test;
+
+public class CommitteeUtilTest {
+
+  @Test
+  void testListShuffleAndShuffledIndexCompatibility() {
+    Bytes32 seed = Bytes32.ZERO;
```
Comment (ajsutton) [inline ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java:26]: Given how comprehensive the reference tests are for shuffling, we can probably delete the original implementation and ditch this test.  *Maybe* hard code one expected shuffle result here just as a quick sanity check as part of unit tests and leave the details to the reference tests.

**Review by ajsutton — APPROVED** [review]
LGTM.

**Comment by ajsutton** [comment]
The unit test failures look quite interesting.  Might just be that those tests are setting up a very invalid state but worth understanding.

**Comment by Nashatyrev** [comment]
> 
> 
> The unit test failures look quite interesting. Might just be that those tests are setting up a very invalid state but worth understanding.

Yep, just a degenerate empty list shuffle case

File: ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java (hunk 2 above)
Comment (Nashatyrev) [reply ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java:26]: The original `compute_shuffled_index` is still actual for computing proposer index. So we should leave it and may leave this simple test

File: eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java (hunk 1 above)
Comment (Nashatyrev) [reply eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java:29]: Here a comparison of java implementations I found: 
https://stackoverflow.com/questions/7846841/how-to-calculate-sha256-faster-in-java

Interesting that Apache implementation is more than 2x faster than Bouncy Castle 

Here is some native implementation wrapped by JCA interface: https://github.com/corretto/amazon-corretto-crypto-provider. Though it is available for Linux only and I couldn't find any benchmark results for it. 

Here I found some Java/Native variants comparison within the same lib, but I'm not sure which Java implementation was used and how good is the native implementation: https://www.nayuki.io/page/native-hash-functions-for-java

Anyways sha256 is far from being a bottleneck for now, so I don't think it worth spending much time for alternatives investigation

File: ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/datastructures/src/test/java/tech/pegasys/artemis/datastructures/util/CommitteeUtilTest.java:26]: ah ok, I thought we'd be able to delete it, so yes testing that they give the same results is definitely the best approach then.  Guess that explains how you got away with only doing backwards search and not forwards...

File: eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java (hunk 1 above)
Comment (ajsutton) [reply eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java:29]: Agreed - purely asking out of interest. :)

Comment (cemozerr) [reply eth-benchmark-tests/src/jmh/java/tech/pegasys/artemis/benchmarks/Sha256Benchmark.java:29]: On a similar vein, what are your thoughts on switching to a faster BLS implementation (I think it was herumi) @Nashatyrev ?

**Pull request merged**